library/request_queue.cpp

```diff
@@ -0,0 +1,288 @@
+	 * Thus any transaction may only be processed on single thread at any given time.
+	 * But it is possible to ping-pong transaction between multiple IO threads as long as each IO thread
+	 * processes different transaction reply simultaneously.
+	 *
+	 * We must set current thread index to -1 to highlight that current thread currently does not perform any task,
+	 * so it can be assigned any transaction reply, if it is not already claimed by another thread.
+	 *
+	 * If we leave here previously processed transaction id, we might stuck, since all threads will wait for those
+	 * transactions they are assigned to, thus not allowing any further process, since no thread will be able to
+	 * process current request and move to the next one.
+	 */
+	wio->trans = ~0ULL;
+
+	if (!list_empty(&wio->reply_list)) {
+		it = list_first_entry(&wio->reply_list, struct dnet_io_req, req_entry);
+		auto cmd = reinterpret_cast<const dnet_cmd *>(it->header);
+		trans = cmd->trans;
+		wio->trans = trans;
+		return it;
+	}
+
+	if (!list_empty(&wio->request_list)) {
+		it = list_first_entry(&wio->request_list, struct dnet_io_req, req_entry);
+		auto cmd = reinterpret_cast<const dnet_cmd *>(it->header);
+		trans = cmd->trans;
+		wio->trans = trans;
+		return it;
+	}
+
+	std::unique_lock<std::mutex> lock(m_locks_mutex);
+
+	list_for_each_entry_safe(it, tmp, &m_queue, req_entry) {
+		auto cmd = reinterpret_cast<const dnet_cmd *>(it->header);
+
+		/* This is not a transaction reply, process it right now */
+		if (!(cmd->flags & DNET_FLAGS_REPLY)) {
+			if (cmd->flags & DNET_FLAGS_NOLOCK)
+				return it;
+
+			locked_keys_t::iterator it_lock;
+			bool inserted;
+			std::tie(it_lock, inserted) = m_locked_keys.insert({cmd->id, reinterpret_cast<dnet_locks_entry *>(nullptr)});
+			if (inserted) {
+				auto lock_entry = take_lock_entry(wio);
+				it_lock->second = lock_entry;
+				return it;
+			} else {
+				auto lock_entry = it_lock->second;
+				dnet_work_io *owner = lock_entry->owner;
+				/* if key is already locked by other pool thread, then move it to request_list of this thread */
+				if (owner) {
+					list_move_tail(&it->req_entry, &owner->request_list);
+				}
+			}
+		} else {
+			trans = cmd->trans;
+			bool trans_in_process = false;
+
+			for (int i = 0; i < pool->num; ++i) {
+				/* Someone claimed transaction @tid */
+				if (pool->wio_list[i].trans == trans) {
+					list_move_tail(&it->req_entry, &pool->wio_list[i].reply_list);
+					trans_in_process = true;
+					break;
+				}
+			}
+
+			if (!trans_in_process) {
+				wio->trans = trans;
+				return it;
+			}
+		}
+	}
+
+	return nullptr;
+}
+
+void dnet_request_queue::release_request(const dnet_io_req *req)
+{
+	auto cmd = reinterpret_cast<const dnet_cmd *>(req->header);
+	if (!(cmd->flags & DNET_FLAGS_REPLY) &&
+	    !(cmd->flags & DNET_FLAGS_NOLOCK)) {
+		release_key(&cmd->id);
+	}
+}
+
+void dnet_request_queue::lock_key(const dnet_id *id)
+{
+	std::unique_lock<std::mutex> lock(m_locks_mutex);
+	while (1) {
```

Comment:
Does it mean, that lock can loop forever and miss `need_exit` flag set by signal for example?